src/hepattn/models/decoder.py

```diff
@@ -292,20 +292,16 @@ def forward(
             query_posenc: Optional query positional encoding.
             key_posenc: Optional key positional encoding.
             attn_mask_transpose: Optional transposed attention mask.
-
-        Returns:
-            Tuple of updated query and key/value embeddings.
         """
-        if query_posenc is not None:
-            q = q + query_posenc
-        if key_posenc is not None:
-            kv = kv + key_posenc
-
-        # Update query/object embeddings with the key/constituent embeddings
-        q = self.q_ca(q, kv=kv, attn_mask=attn_mask, q_mask=q_mask, kv_mask=kv_mask)
-        q = self.q_sa(q, q_mask=q_mask)
+        q_pe = q if query_posenc is None else q + query_posenc
+        kv_pe = kv if key_posenc is None else kv + key_posenc
+
+        q = q + self.q_ca(q_pe, kv=kv_pe, v=kv, attn_mask=attn_mask, q_mask=q_mask, kv_mask=kv_mask)
         q = self.q_dense(q)
 
+        q_pe = q if query_posenc is None else q + query_posenc
+        q = q + self.q_sa(q_pe, kv=q_pe, v=q, q_mask=q_mask)
```

Comment:
Worth trying not using the PEs for the self attention and just using it for cross attention